Log vs throw for host-name resolution when deriving egress subnets from ingress addresses

manadart opened the pull request:
The linked bug describes a scenario where:
- We have acquired a FQDN as a load-balancer address for a CAAS service.
- This is returned as the public address of the service in a call to `network-get`.
- There are no explicit egress subnets declared for the model or the relation context of the `network-get` call.
- The service pod/container is not yet available.
- Therefore, when we invoke our fall-back position attempt to resolve the FQDN as an IP and in turn as an egress subnet, we return an error.
- The workload goes into an error state.

This patch unifies the egress subnet resolution and instead of throwing out all errors, logs a warning when we detect `net.DNSError`, returning a nil slice for egress subnets.

There are additional modularisations following on from #12309 that continue to make `NetworkInfo` more cogent.

## QA steps

Regression:

- Bootstrap to AWS.
- `juju add-space space-1 172.31.80.0/20`.
- `juju deploy percona-cluster --bind "space-1"`.
- Once quiesced, `juju run --unit percona-cluster/0 'network-get cluster'`.
- Execution will succeed and return something akin to:
```
bind-addresses:
- macaddress: 12:08:86:e8:3c:8d
  interfacename: ens5
  addresses:
  - hostname: ""
    address: 172.31.81.88
    cidr: 172.31.80.0/20
- macaddress: 02:d4:9d:5c:e2:ed
  interfacename: fan-252
  addresses:
  - hostname: ""
    address: 252.81.88.1
    cidr: 252.80.0.0/12
egress-subnets:
- 172.31.81.88/32
ingress-addresses:
- 172.31.81.88
- 252.81.88.1
```
This Feature:

- Bootstrap to AWS.
- Save [this](https://raw.githubusercontent.com/charmed-kubernetes/bundle/master/overlays/aws-overlay.yaml) as aws-integrator.yaml
- `juju deploy charmed-kubernetes --overlay aws-integrator.yaml --trust`.
- `juju scp kubernetes-master/0:/home/ubuntu/config ~/.kube/config`.
- Add the k8s as a cloud to the same controller: `juju juju add-k8s k8s`.
- `juju add-model qa k8s/us-east-1 --config logging-config="<root>=DEBUG"`.
- `juju deploy cs:ambassador`.
- `juju config ambassador juju-external-hostname=<app service address>`.
- `juju run --unit ambassador/0 "network-get ambassador"`

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1901749

manadart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [CAAS unit returns load-balancer address for egress in network-get](https://trello.com/c/dUaNT2eM/3147-caas-unit-returns-load-balancer-address-for-egress-in-network-get)

ycliuhw commented on apiserver/facades/agent/uniter/networkinfo.go:
```diff
@@ -165,6 +138,60 @@ func (n *NetworkInfoBase) maybeGetUnitAddress(rel *state.Relation) (corenetwork.
+	}
+
+	var egress []string
+	retryArg := n.retryFactory()
```
```golang
var defaultRetryFactory = func() retry.CallArgs {
	return retry.CallArgs{
		Clock:       clock.WallClock,
		Delay:       3 * time.Second,
                 // 30s is unfortunately not enough for AWS LB to be fully provisioned.
                  // It could take a few mins.
                  // We probably should not retry, ignore `DNSError` and just return empty egress to let 
                  // client retry later.
		MaxDuration: 30 * time.Second, 
	}
}
```

ycliuhw reviewed (commented):
Just did a QA.
Because 30s retry period is not enough. 
The `config-changed` hook still failed.

wallyworld commented:
> Just did a QA.
> Because 30s retry period is not enough.
> The `config-changed` hook still failed.

I think we need to do "best effort" to resolve the egress info, but leave it empty without error if we can't. The charm can always run network-get later to get the required info. Even 30s seems too long to delay the enter scope workflow.

manadart commented:
@wallyworld @ycliuhw I removed the retry and added a check for `net.DNSError` instead. I can't seem to get into the situation described in the bug though. I always seem to be getting a resolution for the host-name.

wallyworld commented:
I pushed commit e001ea4 to add an option to network-get to allow the charm to ask that ingress addresses which are hostnames not be resolved to IP addresses. The default behaviour is as we have now. We still resolve the binding address hostnames (as per the comment about bug 1721368 in the code). But it is wrong to resolve an AWS ELB FQDN to an IP address - the IP address can change, and there's actually more than 1 and which one you get depends on whether LookupHost or ResolveIPAddr is used.

```
$ nslookup a5b0078df5805427c8114d0d436b8817-1507491215.ap-southeast-2.elb.amazonaws.com
Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   a5b0078df5805427c8114d0d436b8817-1507491215.ap-southeast-2.elb.amazonaws.com
Address: 54.153.167.41
Name:   a5b0078df5805427c8114d0d436b8817-1507491215.ap-southeast-2.elb.amazonaws.com
Address: 3.25.31.237
Name:   a5b0078df5805427c8114d0d436b8817-1507491215.ap-southeast-2.elb.amazonaws.com
Address: 13.55.126.148

```

We need to sort of the LookupHost or ResolveIPAddr issue next. This is illustrated when doing QA on this branch. The controller uses ResolveIPAddr to get the ELB IP address and sets <ip>/32 as the egress info. Then network-get uses LookupHost for the ingress info and this results in 2 totally different IP addresses being printed, ostensibly sourced from the same service FQDN.

```
$ juju run --unit ambassador/0 "network-get ambassador"
bind-addresses:
- macaddress: ""
  interfacename: ""
  addresses:
  - hostname: ""
    address: 10.1.38.6
    cidr: ""
egress-subnets:
- 3.25.31.237/32
ingress-addresses:
- 13.55.126.148
```

wallyworld commented:
Also, there's no need to manually add (or later specify) storage for the k8s model.
Juju understands that the underlying cloud is AWS and "Does The Right Thing".

wallyworld commented:
With the egress CIDRs, you actually potentially get different ones (due to the hostname lookup of the ELB FQDN) each time you run network-get

```
$ juju run --unit ambassador/0 "network-get ambassador --resolve-ingress-addresses=false"
bind-addresses:
- macaddress: ""
  interfacename: ""
  addresses:
  - hostname: ""
    address: 10.1.36.4
    cidr: ""
egress-subnets:
- 52.63.110.237/32
ingress-addresses:
- a8f199c0bc4f04bd58e67fac6f114648-1951140932.ap-southeast-2.elb.amazonaws.com

$ juju run --unit ambassador/0 "network-get ambassador --resolve-ingress-addresses=false"
bind-addresses:
- macaddress: ""
  interfacename: ""
  addresses:
  - hostname: ""
    address: 10.1.36.4
    cidr: ""
egress-subnets:
- 13.210.56.7/32
ingress-addresses:
- a8f199c0bc4f04bd58e67fac6f114648-1951140932.ap-southeast-2.elb.amazonaws.com
```

wallyworld commented:
Based on doing QA, I think there's still a change we want to make. If we attempt to calculate the network info before the ELB has finished provisioning, we no longer error which is good. But, the ingress address is set to the unit private address which is wrong and which will later change to the public address. A minimal change to at least improve things (but not fully fix) would be to supply any FQDN as the ingress address and not fall back to private/cloud local address at all if the host name resolution fails.

ycliuhw commented:
This change should fix the config-changed hook failure now. 
But as @wallyworld mentioned, we should not fall back to the private address when the public address that we were looking for was missing. We might need to change the network address filtering logic to make it more restricted.

wallyworld commented:
![](https://github.trello.services/images/mini-trello-icon.png) [network-get hostname resolution](https://trello.com/c/psNe1bQ4/1289-network-get-hostname-resolution)

ycliuhw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [1901749 Juju fails trying to contact AWS ELB](https://trello.com/c/prQLzj5t/1273-1901749-juju-fails-trying-to-contact-aws-elb)

manadart commented:
> Also, there's no need to manually add (or later specify) storage for the k8s model.
> Juju understands that the underlying cloud is AWS and "Does The Right Thing".

I was not able to add the cloud without specifying storage.

wallyworld reviewed (approved):
As discussed live, land this now and follow up with a patch to:
- if we have hostname, that's a valid public address, do not fallback to private
- do not use hostname to create a /32 egress cidr, just leave egress info out if we can't sensibly make a call

manadart commented:
$$merge$$

manadart commented:
Hit the max test duration:
```
panic: test timed out after 30m0s
```
$$merge$$

manadart commented:
Changed the QA steps; indeed, using the overlay and adding the cloud to the controller only, I did not need to specify storage:
- Invocation:
```
$ juju run --unit ambassador/0 "network-get ambassador"
bind-addresses:
- macaddress: ""
  interfacename: ""
  addresses:
  - hostname: ""
    address: 10.1.14.3
    cidr: ""
```
- Workload log:
```
application-ambassador: 11:03:32 DEBUG jujuc running hook tool "network-get" for ambassador/0-juju-run-3481863798087399504
application-ambassador: 11:03:32 ERROR jujuc resolving "a82f94c726aa445fb8d9756572ad7f29-1358302660.us-east-1.elb.amazonaws.com": lookup a82f94c726aa445fb8d9756572ad7f29-1358302660.us-east-1.elb.amazonaws.com on 10.152.183.162:53: no such host
```
- Controller log:
```
machine-0: 11:03:32 WARNING juju.apiserver.uniter unable to determine egress subnet for "a82f94c726aa445fb8d9756572ad7f29-1358302660.us-east-1.elb.amazonaws.com": lookup a82f94c726aa445fb8d9756572ad7f29-1358302660.us-east-1.elb.amazonaws.com on 127.0.0.53:53: no such host
```
